File: lib/path/event_marker.dart
import 'package:pathplanner/commands/command.dart';

class EventMarker {
  String name;
  num waypointRelativePos;
  num? endWaypointRelativePos;
  Command? command;

  EventMarker({
    this.name = '',
    this.waypointRelativePos = 0,
    this.endWaypointRelativePos,
    this.command,
  });

  EventMarker.fromJson(Map<String, dynamic> json)
      : this(
            name: json['name'],
            waypointRelativePos: json['waypointRelativePos'],
            endWaypointRelativePos: json['endWaypointRelativePos'],
            command: json['command'] != null
                ? Command.fromJson(json['command'])
                : null);

  bool get isZoned => endWaypointRelativePos != null;

  Map<String, dynamic> toJson() {
    return {
      'name': name,
      'waypointRelativePos': waypointRelativePos,
      'endWaypointRelativePos': endWaypointRelativePos,
      'command': command?.toJson(),
    };
  }

  EventMarker clone() {
    return EventMarker(
      name: name,
      waypointRelativePos: waypointRelativePos,
      endWaypointRelativePos: endWaypointRelativePos,
      command: command?.clone(),
    );
  }

  @override
  bool operator ==(Object other) =>
      other is EventMarker &&
      other.runtimeType == runtimeType &&
      other.name == name &&
      other.waypointRelativePos == waypointRelativePos &&
      other.endWaypointRelativePos == endWaypointRelativePos &&
      other.command == command;

  @override
  int get hashCode =>
      Object.hash(name, waypointRelativePos, endWaypointRelativePos, command);
}

File: pathplannerlib/src/main/native/cpp/pathplanner/lib/path/EventMarker.cpp
#include "pathplanner/lib/path/EventMarker.h"
#include "pathplanner/lib/auto/CommandUtil.h"

using namespace pathplanner;

EventMarker EventMarker::fromJson(const wpi::util::json &json) {
	std::string name = json.at("name").get_string();
	double pos = json.at("waypointRelativePos").get_number();
	double endPos = -1.0;
	if (json.contains("endWaypointRelativePos")
			&& !json.at("endWaypointRelativePos").is_null()) {
		endPos = json.at("endWaypointRelativePos").get_number();
	}

	if (!json.at("command").is_null()) {
		return EventMarker(name, pos, endPos,
				CommandUtil::commandFromJson(json.at("command"), false, false));
	}

	return EventMarker(name, pos, endPos, wpi::cmd::None());
}
